api/maven-api-core/src/main/java/org/apache/maven/api/Project.java

```diff
@@ -234,4 +239,32 @@ default String getId() {
      */
     @Nonnull
     Optional<Project> getParent();
+
+    /**
+     * {@return all source root directories}, including the disabled ones, for all languages and scopes.
+     * For listing only the {@linkplain SourceRoot#enabled() enabled} source roots,
+     * the following code can be used:
+     *
+     * <pre>{@literal
+     * List<SourceRoot> enabledRoots = project.getSourceRoots()
+     *         .stream().filter(SourceRoot::enabled).toList();
+     * }</pre>
+     *
+     * The iteration order is the order in which the sources are declared in the POM file.
+     */
+    @Nonnull
+    Collection<SourceRoot> getSourceRoots();
+
+    /**
+     * {@return all enabled sources that provide files in the given language for the given scope}.
+     * If the given scope is {@code null}, then this method returns the enabled sources for all scopes.
+     * If the given language is {@code null}, then this method returns the enabled sources for all languages.
+     * An arbitrary number of source roots may exist for the same scope and language.
+     * It may be, for example, the case of a multi-versions project.
+     * The iteration order is the order in which the sources are declared in the POM file.
+     *
+     * @param scope the scope of the sources to return, or {@code null} for all scopes
+     * @param language the language of the sources to return, or {@code null} for all languages
+     */
+    Stream<SourceRoot> getEnabledSourceRoots(ProjectScope scope, Language language);
```

Comment:
The original idea was to have the `Project` completely immutable and have modifications done through the `ProjectManager` service.  Plugins can add sources during the build, so both accessors and setters are currently defined on `ProjectManager`, see https://github.com/apache/maven/blob/dc6d48ce00c3ebc0579079cadc93ae6175ed183b/api/maven-api-core/src/main/java/org/apache/maven/api/services/ProjectManager.java#L120-L158

Reply:
The methods to remove would be `Project.getSourceRoots()` and `getEnabledSourceRoots(…)`? I thought that the [previous comment](https://github.com/apache/maven/pull/2061#discussion_r1930340516) was about adding them.

Reply:
Yes, but we need methods to add a `Source` to a given project at runtime I think.
The incoherence is more about the location of these.  The `Project` is supposed to be immutable, that's why they are located inside the `ProjectManager` which purpose is to hold modifications to the project.

Reply:
Pushed addition of two `addSourceRoot` method (the second one is a convenience) and removed the 4 above-cited methods.

Note: for interfaces such as `Project` that are intended to be unmodifiable, should be remove the `get` prefix? It would express the intent that those interfaces have no setter methods, and would allow to implement some of them with `record`. The latter may not be applicable to `DefaultProject`, but some other implementations may benefit from that.

Reply:
Yes, it could make sense to align with the `record` way of doing things.

Reply:
I guess the most important ones would be the model classes themselves, but this would be quite a big change.

Reply:
Maybe we should start with interfaces that are new in Maven 4. Nevertheless closing this conversation as it would be the topic for a separated pull request.